ir/src/lib.rs

```diff
@@ -69,6 +76,8 @@ impl AirIR {
                     symbol_table.insert_constants(constants)?;
                 }
                 ast::SourceSection::TraceCols(columns) => {
+                    // set the overall trace columns count
+                    num_polys = (columns.main_cols.len() as u8, columns.aux_cols.len() as u8);
```

Comment:
This won't quite work, because we could have more than one `TraceCols` source section (e.g. once we support modules).

Instead of doing it here and initializing above, maybe we should collect this information inside the symbol table, the way that we collect the `num_trace_segments` information. Then you could update the `into_declarations` method to pass the information you need to initialize `num_polys`. I think this will encapsulate things a bit more cleanly.